src/backend/distributed/utils/background_jobs.c

```diff
@@ -385,292 +648,339 @@ CitusBackgroundTaskQueueMonitorMain(Datum arg)
-		MemoryContextSwitchTo(oldContext);
-
-		if (!task || task->status == BACKGROUND_TASK_STATUS_CANCELLING ||
-			task->status == BACKGROUND_TASK_STATUS_CANCELLED)
+		BackgroundExecutorHashEntry *handleEntry = NULL;
+		foreach_ptr(handleEntry, runningTaskEntries)
 		{
-			/*
-			 * A concurrent cancel has happened or the task has disappeared, we are not
-			 * retrying or changing state, we will only reflect the message onto the task
-			 * and for completeness we update the job aswell, this should be a no-op
-			 *
-			 * We still need to release the shared memory and xact before looping
-			 */
-
-			dsm_detach(seg);
+			/* execution state for current task in FSM */
+			BackgroundMonitorExecutionStates monitorExecutionState = ExecutionStarted;
 
-			task->status = BACKGROUND_TASK_STATUS_CANCELLED;
-			task->message = message.data;
-			UpdateBackgroundTask(task);
-			UpdateBackgroundJob(task->jobid);
+			/* useful to store cancelling status for current task in FSM */
+			bool taskCancelledConcurrently = false;
 
-			PopActiveSnapshot();
-			CommitTransactionCommand();
+			/* currently running task used in FSM */
+			BackgroundTask *task = NULL;
 
-			continue;
-		}
-		else if (hadError)
-		{
-			/*
-			 * When we had an error we need to decide if we want to retry (keep the
-			 * runnable state), or move to error state
-			 */
-			if (!task->retry_count)
+			while (monitorExecutionState != ExecutionEnded)
 			{
-				SET_NULLABLE_FIELD(task, retry_count, 1);
-			}
-			else
-			{
-				(*task->retry_count)++;
+				switch (monitorExecutionState)
+				{
+					case ExecutionStarted:
+					{
+						/*
+						 * initial execution state
+						 * check if we started execution after a delay
+						 */
+
+						if (queueMonitorExecutionContext.backgroundWorkerFailedStartTime >
+							0)
+						{
+							/*
+							 * we had a delay in starting the background worker for task execution. Report
+							 * the actual delay and reset the time. This allows a subsequent task to
+							 * report again if it can't start a background worker directly.
+							 */
+							long secs = 0;
+							int microsecs = 0;
+							TimestampDifference(
+								queueMonitorExecutionContext.
+								backgroundWorkerFailedStartTime,
+								GetCurrentTimestamp(),
+								&secs, &microsecs);
+							ereport(LOG, (errmsg(
+											  "able to start a background worker with %ld seconds"
+											  "delay", secs)));
+
+							queueMonitorExecutionContext.backgroundWorkerFailedStartTime =
+								0;
+						}
+
+						monitorExecutionState = TaskConcurrentCancelCheck;
+						break;
+					}
+
+					case TaskConcurrentCancelCheck:
+					{
+						/*
+						 * here we take exclusive lock on pg_dist_background_task table to prevent a
+						 * concurrent modification. We first check if a task is concurrently cancelled
+						 * or deleted. Accordingly, we decide to make the task status as cancelled or running.
+						 */
+
+						StartTransactionCommand();
```

Comment:
I find it slightly hard to know for sure that we eventually commit this transaction in this loop. Is there a different way to do this? e.g. we always open a transaction?

Reply:
we can actually open transaction at the start of the loop and commit it at the end to make it more readable. Thats what we do in the loop but unfriendly way.